README.md
# NewebPay Implementation

[![CircleCI](https://circleci.com/gh/libterty/newebpay/tree/master.svg?style=svg)](https://circleci.com/gh/libterty/newebpay/tree/master)

- Referencing NewebPay API Document to perform API calling and bundle it into a module

## How to install

```
npm install newebpay
```

## How to use

- Declare for the module globally

```javascript
const { TradeModules } = require('newebpay');
```

- Initialize module

```javascript
const trade = new TradeModules(
  'your domain information <string>',
  'your newebpay MerchantID <string>',
  'your newebpay HashKey <string>',
  'your newebpay HashIV <string>',
  'newebpay PayGateWay <string>',
  'your newebpay ClientBackURL <string>',
  'your ReturnURL <string>',
  'your NotifyURL <string>',
);
```

## CallIng API

- When you want to return tradeInfo then calling Newebpay Authorization

```javascript
const tradeInfo = trade.getTradeInfo(
  (Amt: number),
  (DESC: string),
  (email: string),
  (comment: string),
);
```

- Expecting `tradeInfo` output example

```json
{
  "MerchantID": "your newebpay MerchantID <string>",
  "TradeInfo": "aes256 encryption",
  "TradeSha": "sha256 hashing",
  "Version": "1.5",
  "PayGateWay": "https://ccore.spgateway.com/MPG/mpg_gateway",
  "MerchantOrderNo": "1581686501236"
}
```

- When you want Decrypt Newebpay Feedback Information

```javascript
const tradeData = JSON.parse(trade.createMpgAesDecrypt(TradeInfo));
```

# ChangeLog

[ChangeLog](https://github.com/libterty/newebpay/blob/master/ChangeLog.md)

# CopyRight

Copyright © 2020, 11. Released under the [MIT License](https://github.com/libterty/newebpay/blob/master/LICENCE).


ChangeLog.md
## Version 0.0.11 (2021-04-06)

## Updating Document

- Update constractor and getTradeInfo

## Version 0.0.9 (2019-02-15)

## New features

## Version 0.0.10 (2019-02-15)

## Updating Document

- Revise Readme Decrypt comment Info[``]()

## Version 0.0.9 (2019-02-15)

## New features

- Reducing Build file[`1e17889d8c58e614baa3e3d74155bcc99e70cf0c`](https://github.com/libterty/newebpay/commit/1e17889d8c58e614baa3e3d74155bcc99e70cf0c)

## Version 0.0.8 (2019-02-15)

## New features

- Adding Unit Test[`3c43cd2ea850d97ef3a03a0c88250d5a71489a3d`](https://github.com/libterty/newebpay/commit/3c43cd2ea850d97ef3a03a0c88250d5a71489a3d)

### Bug Fixes

- Fixing getTradeInfo thrown issue[`3c43cd2ea850d97ef3a03a0c88250d5a71489a3d`](https://github.com/libterty/newebpay/commit/3c43cd2ea850d97ef3a03a0c88250d5a71489a3d)
- Fixing function scope[`3c43cd2ea850d97ef3a03a0c88250d5a71489a3d`](https://github.com/libterty/newebpay/commit/3c43cd2ea850d97ef3a03a0c88250d5a71489a3d)

### Updating Document

- Updating Readme and ChangeLog[`3c43cd2ea850d97ef3a03a0c88250d5a71489a3d`](https://github.com/libterty/newebpay/commit/3c43cd2ea850d97ef3a03a0c88250d5a71489a3d)

## Version 0.0.7 (2019-02-14)

### Updating Document

- Updating Readme and ChangeLog[`263fd3be7b21b9310c2571ce94aa6e4e6ca60ce0`](https://github.com/libterty/newebpay/commit/263fd3be7b21b9310c2571ce94aa6e4e6ca60ce0)

## Version 0.0.6 (2019-02-14)

### Bug Fixes

- Fixing Email Check Err[`b60d79ac667c7375f08f9430f0ab4089c84685c6`](https://github.com/libterty/newebpay/commit/b60d79ac667c7375f08f9430f0ab4089c84685c6)

## Version 0.0.5 (2019-02-14)

### Bug Fixes

- Updating Package Version[`a5507b7da6e0133a62a8ce3e79cae1d8ec3fbfc0`](https://github.com/libterty/newebpay/commit/a5507b7da6e0133a62a8ce3e79cae1d8ec3fbfc0)

## Version 0.0.4 (2019-02-14)

### New features

- Update getTradeInfo with email check[`d280f5806ebd8293cf79e0bb85d1cd471990f17d`](https://github.com/libterty/newebpay/commit/d280f5806ebd8293cf79e0bb85d1cd471990f17d)

## Version 0.0.3 (2019-02-14)

### New features

- Updating npm configuration[`9db4af7251f547840f0b3f1d1a9346ed2ab4ea41`](https://github.com/libterty/NewebPay/commit/9db4af7251f547840f0b3f1d1a9346ed2ab4ea41)
- Updating package name[`a413c82ac500d456d8952f7f763d03991fb42737`](https://github.com/libterty/NewebPay/commit/a413c82ac500d456d8952f7f763d03991fb42737)

## Version 0.0.2 (2019-02-14)

### New features

- Adding trading algo[`492da1221c961ccc19b2dfacc35a74a89db08eca`](https://github.com/libterty/NewebPay/commit/492da1221c961ccc19b2dfacc35a74a89db08eca)
- Adding Url Check[`8e456b2d1b8bc53b9d6de4fd9f9bd7aa1381deb5`](https://github.com/libterty/NewebPay/commit/8e456b2d1b8bc53b9d6de4fd9f9bd7aa1381deb5)

## Version 0.0.1 (2019-02-14)

### New features

- project init[`17f7c716864de254b9e66ea63f493bafdc6f4be9`](https://github.com/libterty/NewebPay/commit/17f7c716864de254b9e66ea63f493bafdc6f4be9)
- setting for npm publish[`9d146c2c69669f38eb482a76d18ae3fb6ffe1ad0`](https://github.com/libterty/NewebPay/commit/9d146c2c69669f38eb482a76d18ae3fb6ffe1ad0)
- Updating Package.json Repo[`6ffb70123b87421a70729140b3758ba3a121fd95`](https://github.com/libterty/NewebPay/commit/6ffb70123b87421a70729140b3758ba3a121fd95)


src/index.ts
import { createCipheriv, createDecipheriv, createHash } from 'crypto';
import { IsNotEmpty, IsUrl } from 'class-validator';

export class TradeModules {
  @IsNotEmpty()
  @IsUrl()
  URL: string;
  @IsNotEmpty()
  MerchantID: string;
  @IsNotEmpty()
  HashKey: string;
  @IsNotEmpty()
  HashIV: string;
  @IsNotEmpty()
  @IsUrl()
  PayGateWay: string;
  @IsNotEmpty()
  @IsUrl()
  ReturnURL: string;
  @IsNotEmpty()
  @IsUrl()
  NotifyURL: string;
  @IsNotEmpty()
  @IsUrl()
  ClientBackURL: string;

  constructor(
    url: string,
    merchantId: string,
    hashKey: string,
    hashIV: string,
    payGateWay: string,
    clientBackURL: string,
    ReturnURL?: string,
    NotifyURL?: string,
  ) {
    this.URL = url;
    this.MerchantID = merchantId;
    this.HashKey = hashKey;
    this.HashIV = hashIV;
    this.PayGateWay = payGateWay;
    this.ReturnURL =
      ReturnURL || this.URL + '/spgateway/callback?from=ReturnURL';
    this.NotifyURL =
      NotifyURL || this.URL + '/spgateway/callback?from=NotifyURL';
    this.ClientBackURL = clientBackURL || 'http://localhost:8080/orders';
  }

  private genDataChain(TradeInfo: object): string {
    const results = [];
    for (const kv of Object.entries(TradeInfo)) {
      results.push(`${kv[0]}=${kv[1]}`);
    }
    return results.join('&');
  }

  private createMpgAesEncrypt(TradeInfo: object): string {
    const encrypt = createCipheriv('aes256', this.HashKey, this.HashIV);
    const enc = encrypt.update(this.genDataChain(TradeInfo), 'utf8', 'hex');
    return enc + encrypt.final('hex');
  }

  public createMpgAesDecrypt(TradeInfo: string): string {
    const decrypt = createDecipheriv('aes256', this.HashKey, this.HashIV);
    decrypt.setAutoPadding(false);
    const text = decrypt.update(TradeInfo, 'hex', 'utf8');
    const plainText = text + decrypt.final('utf8');
    const result = plainText.replace(/[\x00-\x20]+/g, '');
    return result;
  }

  private createMpgShaEncrypt(TradeInfo: string): string {
    const sha = createHash('sha256');
    const plainText = `HashKey=${this.HashKey}&${TradeInfo}&HashIV=${this.HashIV}`;

    return sha
      .update(plainText)
      .digest('hex')
      .toUpperCase();
  }

  public getTradeInfo(
    Amt: number,
    Desc: string,
    email: string,
    comment?: string,
  ): any {
    const emailRule = /^\w+((-\w+)|(\.\w+))*\@[A-Za-z0-9]+((\.|-)[A-Za-z0-9]+)*\.[A-Za-z]+$/;

    if (Amt < 0)
      return { status: 'error', message: `Input Amt type ${Amt} Error` };
    if (Desc.length < 1)
      return {
        status: 'error',
        message: `Input Desc Length ${Desc} Error`,
      };
    if (email.search(emailRule) !== -1) {
      const data = {
        MerchantID: this.MerchantID,
        RespondType: 'JSON',
        TimeStamp: Date.now(),
        Version: 1.5,
        MerchantOrderNo: Date.now(),
        LoginType: 0,
        OrderComment: comment || 'OrderComment',
        Amt: Amt,
        ItemDesc: Desc,
        Email: email,
        ReturnURL: this.ReturnURL,
        NotifyURL: this.NotifyURL,
        ClientBackURL: this.ClientBackURL,
      };
      const mpgAesEncrypt = this.createMpgAesEncrypt(data);
      const mpgShaEncrypt = this.createMpgShaEncrypt(mpgAesEncrypt);
      const tradeInfo = {
        MerchantID: this.MerchantID,
        TradeInfo: mpgAesEncrypt,
        TradeSha: mpgShaEncrypt,
        Version: 1.5,
        PayGateWay: this.PayGateWay,
        MerchantOrderNo: data.MerchantOrderNo,
      };
      return tradeInfo;
    } else {
      return {
        status: 'error',
        message: `Input Email content ${email} Error`,
      };
    }
  }
}


lib/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var crypto_1 = require("crypto");
var class_validator_1 = require("class-validator");
var TradeModules = /** @class */ (function () {
    function TradeModules(url, merchantId, hashKey, hashIV, payGateWay, clientBackURL, ReturnURL, NotifyURL) {
        this.URL = url;
        this.MerchantID = merchantId;
        this.HashKey = hashKey;
        this.HashIV = hashIV;
        this.PayGateWay = payGateWay;
        this.ReturnURL =
            ReturnURL || this.URL + '/spgateway/callback?from=ReturnURL';
        this.NotifyURL =
            NotifyURL || this.URL + '/spgateway/callback?from=NotifyURL';
        this.ClientBackURL = clientBackURL || 'http://localhost:8080/orders';
    }
    TradeModules.prototype.genDataChain = function (TradeInfo) {
        var results = [];
        for (var _i = 0, _a = Object.entries(TradeInfo); _i < _a.length; _i++) {
            var kv = _a[_i];
            results.push(kv[0] + "=" + kv[1]);
        }
        return results.join('&');
    };
    TradeModules.prototype.createMpgAesEncrypt = function (TradeInfo) {
        var encrypt = crypto_1.createCipheriv('aes256', this.HashKey, this.HashIV);
        var enc = encrypt.update(this.genDataChain(TradeInfo), 'utf8', 'hex');
        return enc + encrypt.final('hex');
    };
    TradeModules.prototype.createMpgAesDecrypt = function (TradeInfo) {
        var decrypt = crypto_1.createDecipheriv('aes256', this.HashKey, this.HashIV);
        decrypt.setAutoPadding(false);
        var text = decrypt.update(TradeInfo, 'hex', 'utf8');
        var plainText = text + decrypt.final('utf8');
        var result = plainText.replace(/[\x00-\x20]+/g, '');
        return result;
    };
    TradeModules.prototype.createMpgShaEncrypt = function (TradeInfo) {
        var sha = crypto_1.createHash('sha256');
        var plainText = "HashKey=" + this.HashKey + "&" + TradeInfo + "&HashIV=" + this.HashIV;
        return sha
            .update(plainText)
            .digest('hex')
            .toUpperCase();
    };
    TradeModules.prototype.getTradeInfo = function (Amt, Desc, email, comment) {
        var emailRule = /^\w+((-\w+)|(\.\w+))*\@[A-Za-z0-9]+((\.|-)[A-Za-z0-9]+)*\.[A-Za-z]+$/;
        if (Amt < 0)
            return { status: 'error', message: "Input Amt type " + Amt + " Error" };
        if (Desc.length < 1)
            return {
                status: 'error',
                message: "Input Desc Length " + Desc + " Error",
            };
        if (email.search(emailRule) !== -1) {
            var data = {
                MerchantID: this.MerchantID,
                RespondType: 'JSON',
                TimeStamp: Date.now(),
                Version: 1.5,
                MerchantOrderNo: Date.now(),
                LoginType: 0,
                OrderComment: comment || 'OrderComment',
                Amt: Amt,
                ItemDesc: Desc,
                Email: email,
                ReturnURL: this.ReturnURL,
                NotifyURL: this.NotifyURL,
                ClientBackURL: this.ClientBackURL,
            };
            var mpgAesEncrypt = this.createMpgAesEncrypt(data);
            var mpgShaEncrypt = this.createMpgShaEncrypt(mpgAesEncrypt);
            var tradeInfo = {
                MerchantID: this.MerchantID,
                TradeInfo: mpgAesEncrypt,
                TradeSha: mpgShaEncrypt,
                Version: 1.5,
                PayGateWay: this.PayGateWay,
                MerchantOrderNo: data.MerchantOrderNo,
            };
            return tradeInfo;
        }
        else {
            return {
                status: 'error',
                message: "Input Email content " + email + " Error",
            };
        }
    };
    __decorate([
        class_validator_1.IsNotEmpty(),
        class_validator_1.IsUrl()
    ], TradeModules.prototype, "URL", void 0);
    __decorate([
        class_validator_1.IsNotEmpty()
    ], TradeModules.prototype, "MerchantID", void 0);
    __decorate([
        class_validator_1.IsNotEmpty()
    ], TradeModules.prototype, "HashKey", void 0);
    __decorate([
        class_validator_1.IsNotEmpty()
    ], TradeModules.prototype, "HashIV", void 0);
    __decorate([
        class_validator_1.IsNotEmpty(),
        class_validator_1.IsUrl()
    ], TradeModules.prototype, "PayGateWay", void 0);
    __decorate([
        class_validator_1.IsNotEmpty(),
        class_validator_1.IsUrl()
    ], TradeModules.prototype, "ReturnURL", void 0);
    __decorate([
        class_validator_1.IsNotEmpty(),
        class_validator_1.IsUrl()
    ], TradeModules.prototype, "NotifyURL", void 0);
    __decorate([
        class_validator_1.IsNotEmpty(),
        class_validator_1.IsUrl()
    ], TradeModules.prototype, "ClientBackURL", void 0);
    return TradeModules;
}());
exports.TradeModules = TradeModules;
